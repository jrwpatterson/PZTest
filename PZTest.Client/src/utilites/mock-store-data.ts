import { RootReducer } from '../root-reducer'

export const defaultMockStore: RootReducer = {
  basketReducer: {
    basket: {
      id: 'testID',
      lines: [
        {
          grams: 200,
          name: 'test product',
          productID: 'testProdID',
          rowNo: 0,
          price: 5,
        },
        {
          grams: 400,
          name: 'test product 2',
          productID: 'testProdID2',
          rowNo: 1,
          price: 20,
        },
      ],
      total: 25,
    },
    validBasket: true,
  },
  cheeseReducer: {
    cheeses: [
      {
        cheeseColour: 'purple',
        id: 'testid',
        name: 'testCheese',
        pictureUrl: 'testUrl',
        pricePerKG: 10,
      },
      {
        cheeseColour: 'pink',
        id: 'testid2',
        name: 'testPinkCheese',
        pictureUrl: 'testPinkUrl',
        pricePerKG: 15,
      },
    ],
  },
}
