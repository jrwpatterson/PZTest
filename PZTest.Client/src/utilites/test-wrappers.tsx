import React, { ReactType, ReactNode } from 'react'
import { Provider } from 'react-redux'
import configureStore, { MockStoreEnhanced } from 'redux-mock-store'
import { defaultMockStore } from './mock-store-data'
import { render } from '@testing-library/react'
import { renderHook, RenderHookOptions } from '@testing-library/react-hooks'
import { RootReducer } from '../root-reducer'
// import { createMount } from '@material-ui/core/test-utils'
import { Provider as FetchProvider } from 'use-http'
import { MemoryRouter } from 'react-router-dom'

const mockStore = configureStore([])

const createStore = (storeData: RootReducer) => {
  return mockStore(storeData)
}

export const createMockProvider = (
  storeData: RootReducer = defaultMockStore,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={createStore(storeData)}>{children}</Provider>
    </FetchProvider>
  )
}

export const createMockProviderWithStore = (
  fakeStore: MockStoreEnhanced<unknown, {}>,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={fakeStore}>{children}</Provider>
    </FetchProvider>
  )
}

export const renderReduxConnectedComponent = (
  children: ReactType | JSX.Element,
  storeOverride: RootReducer = defaultMockStore,
  initialEntries?: string[],
) => {
  const storeData = storeOverride || defaultMockStore
  // const mount = createMount()
  const store = createStore(storeData)
  const Wrapper = createMockProviderWithStore(store)
  const element = (
    <MemoryRouter initialEntries={initialEntries}>
      <Wrapper>{children}</Wrapper>
    </MemoryRouter>
  )
  const component = render(element)

  return {
    store,
    ...component,
  }
}

export const renderReduxConnectedHook = <T extends {}>(
  callback: (props: unknown) => T,
  storeData: RootReducer = defaultMockStore,
  options: RenderHookOptions<unknown> = {},
) => {
  const store = createStore(storeData)
  const wrapper = createMockProviderWithStore(store)
  const renderedHook = renderHook(callback, { ...options, wrapper })

  return {
    store,
    ...renderedHook,
  }
}
