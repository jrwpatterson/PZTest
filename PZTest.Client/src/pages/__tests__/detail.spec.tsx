import React from 'react'
import { renderReduxConnectedComponent } from '../../utilites/test-wrappers'
import Detail, { calculatePrice } from '../detail'
import { defaultMockStore } from '../../utilites/mock-store-data'
import { Route } from 'react-router-dom'

describe('details screen', () => {
  it('should match its snapshot', () => {
    const { container } = renderReduxConnectedComponent(
      <Route path='/detail/:id'>
        <Detail />
      </Route>,
      undefined,
      ['/detail/testid2'],
    )
    expect(container).toMatchSnapshot()
  })

  it('should show an invalid product message if no product', () => {
    const { queryByText } = renderReduxConnectedComponent(
      <Route path='/detail/:id'>
        <Detail />
      </Route>,
      {
        ...defaultMockStore,
        cheeseReducer: {
          cheeses: [
            {
              name: 'nothing',
              pricePerKG: 1,
              pictureUrl: 't',
              cheeseColour: 'noo',
              id: 'nothing',
            },
          ],
        },
      },
    )
    expect(queryByText('Invalid Product')).toBeDefined()
  })

  it('should show the details of the product', () => {
    const { getByTestId } = renderReduxConnectedComponent(
      <Route path='/detail/:id'>
        <Detail />
      </Route>,
      undefined,
      ['/detail/testid2'],
    )
    expect(getByTestId('projected.cost').innerHTML).toEqual(
      'Calculated Price $0',
    )
  })

  it('should calculate the price correctly', () => {
    expect(calculatePrice(600, 10)).toBe(6)
  })

  it('should have a slider that calculates the estimated price for basket', () => {
    throw new Error('not implemented')
  })

  it('should have an add to basket button that adds to basket', () => {
    throw new Error('not implemented')
  })
})
