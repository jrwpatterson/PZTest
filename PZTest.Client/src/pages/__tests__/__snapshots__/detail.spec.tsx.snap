// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`details screen should match its snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <h1>
      testPinkCheese
    </h1>
    <div
      class="makeStyles-outerContainer-2"
    >
      <div
        class="makeStyles-detailsContainer-3"
      >
        <img
          alt="testPinkCheese"
          class="makeStyles-image-4"
          src="testPinkUrl"
        />
        <div>
          <span>
            Colour - 
          </span>
          <span>
            pink
          </span>
        </div>
        <div>
          <span>
            Price per kg - 
          </span>
          <span>
            15
          </span>
        </div>
        <div
          class="makeStyles-addToBasketBox-5"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
            id="discrete-slider-small-steps"
          >
            Calculate Price
          </p>
          <span
            class="MuiSlider-root MuiSlider-colorPrimary"
            id="slider"
          >
            <span
              class="MuiSlider-rail"
            />
            <span
              class="MuiSlider-track"
              style="left: 0%; width: 0%;"
            />
            <input
              type="hidden"
              value="0"
            />
            <span
              aria-labelledby="discrete-slider-small-steps"
              aria-orientation="horizontal"
              aria-valuemax="2000"
              aria-valuemin="0"
              aria-valuenow="0"
              class="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-57"
              data-index="0"
              role="slider"
              style="left: 0%;"
              tabindex="0"
            >
              <span
                class="PrivateValueLabel-offset-59 MuiSlider-valueLabel"
              >
                <span
                  class="PrivateValueLabel-circle-60"
                >
                  <span
                    class="PrivateValueLabel-label-61"
                  >
                    0
                  </span>
                </span>
              </span>
            </span>
          </span>
          <div
            class="makeStyles-addToBasketLowerBox-6"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Add to basket
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
              id="discrete-slider-small-steps"
            >
              <span
                data-testid="projected.cost"
              >
                Calculated Price $
                0
              </span>
            </p>
          </div>
        </div>
      </div>
      <div>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                Basket
              </span>
            </div>
          </li>
          <hr
            class="MuiDivider-root"
          />
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                test product
              </span>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                 - 
              </span>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                $5
              </span>
            </div>
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                test product 2
              </span>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                 - 
              </span>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                $20
              </span>
            </div>
          </li>
          <hr
            class="MuiDivider-root"
          />
          <li
            class="MuiListItem-root MuiListItem-gutters"
            data-testid="basket.total"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                Total
              </span>
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                 - 
              </span>
            </div>
            <div
              class="MuiListItemText-root"
              data-testid="basket.total"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
              >
                $25
              </span>
            </div>
          </li>
          <hr
            class="MuiDivider-root"
          />
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Clear Basket
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
